### PR TITLE
Handle 'Node with this Hostname already exists'.

### PR DESCRIPTION
**Pull request by jameinel**
If we have a failure to actually start an LXD (or KVM) container, then
we would mark it as in error, but would retry the provisioning after a
10s delay. However, we should not try to CreateDevice() again in MAAS,
(or we should be Delete()ing the device). The way we were doing it was
causing us to Create the device twice, which was why MAAS was giving an
error the second time.

This was tested on a live MAAS by injecting a failure just before
broker.manager.CreateContainer. Which should cause us to fail to
provision the container with that specific error, rather than failing
with Hostname already exists.

## QA steps

You need some way to cause CreateContainer to fail. In my testing I used:
```
diff --git a/worker/provisioner/lxd-broker.go b/worker/provisioner/lxd-broker.go
index c8cc50f..749e378 100644
--- a/worker/provisioner/lxd-broker.go
+++ b/worker/provisioner/lxd-broker.go
@@ -120,6 +120,7 @@ func (broker *lxdBroker) StartInstance(args environs.StartInstanceParams) (*envi
        }

        storageConfig := &container.StorageConfig{}
+       return nil, errors.Errorf("boom, lxd shall not pass")
        inst, hardware, err := broker.manager.CreateContainer(
                args.InstanceConfig, args.Constraints,
                series, network, storageConfig, args.StatusCallback,
```

At which point you should be able to do:
```
  $ juju bootstrap maas
  $ juju switch controller
  $ juju add-machine lxd:0
```
And it should fail to provision the container, with the error "boom, lxd shall not pass", rather than the error "Node with this Hostname already exists".

## Documentation changes

No.

## Bug reference

[lp:1670873](https://bugs.launchpad.net/juju/+bug/1670873)

### CODE REVIEW

File: provider/maas/environ.go [hunk 1]
```diff
@@ -2246,15 +2246,39 @@ func (env *maasEnviron) allocateContainerAddresses2(hostInstanceID instance.Id,
...
+	maybeDevices, err := machine.Devices(devicesArgs)
 	if err != nil {
-		return nil, errors.Trace(err)
+		logger.Warningf("error while trying to lookup %q: %v", deviceName, err)
+	} else {
+		if len(maybeDevices) == 1 {
+			logger.Debugf("found MAAS device for container %q, using existing device", deviceName)
+			device = maybeDevices[0]
+		} else if len(maybeDevices) > 1 {
+			logger.Warningf("found too many MAAS devices (%d) for container %q", len(maybeDevices), deviceName)
```
Comment (perrito666) [inline provider/maas/environ.go]: expected N got N+M is better as a logging messages for the eye of the untrained log skimmer.

File: provider/maas/environ.go [hunk 2]
```diff
@@ -2246,15 +2246,39 @@ func (env *maasEnviron) allocateContainerAddresses2(hostInstanceID instance.Id,
...
+		// The device didn't already exist, so we Create it.
+		createDeviceArgs := gomaasapi.CreateMachineDeviceArgs{
+			Hostname:      deviceName,
+			MACAddress:    primaryMACAddress,
+			Subnet:        subnet, // can be nil
+			InterfaceName: primaryNICName,
+		}
+		device, err = machine.CreateDevice(createDeviceArgs)
+		if err != nil {
+			// TODO(jam): I believe MAAS is returning "Hostname already exists" if
```
Comment (perrito666) [inline provider/maas/environ.go]: This todo could benefit from a bug number.

Comment (jameinel) [inline provider/maas/environ.go]: The TODO no longer applies because I am doing the lookup first always.

File: provider/maas/environ.go (hunk 1 above)
Comment (jameinel) [inline provider/maas/environ.go]: logger.Warningf("found more than 1 MAAS devices (%d) for container %q", len(maybeDevices), deviceName)

better?

**Comment by jameinel** [comment]
$$merge$$

**Comment by jameinel** [comment]
$$merge$$ added an error path but had missed the return needed 2 parameters.

**Pull request merged**